src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,290 @@
-                // check whether we are good to report total symbol numbers
-                if (_symbolCountByProjectMap == null || _symbolCountByProjectMap.Count < solution.ProjectIds.Count || string.IsNullOrEmpty(solution.FilePath))
+                if (!document.SupportsSyntaxTree)
                 {
+                    // Not a language we can produce indices for (i.e. TypeScript).  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                if (_solutionId != null && _solutionId != solution.Id)
+                if (bodyOpt != null)
                 {
-                    ReportCount();
+                    // This was a method level edit.  This can't change the symbol tree info
+                    // for this project.  Bail immediately.
                     return SpecializedTasks.EmptyTask;
                 }
 
-                _solutionId = solution.Id;
-                foreach (var projectId in solution.ProjectIds)
+                return UpdateSymbolTreeInfoAsync(document.Project, cancellationToken);
+            }
+
+            public override Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
+            {
+                return UpdateSymbolTreeInfoAsync(project, cancellationToken);
+            }
+
+            private async Task UpdateSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
                 {
-                    if (!_symbolCountByProjectMap.ContainsKey(projectId))
-                    {
-                        return SpecializedTasks.EmptyTask;
-                    }
+                    return;
+                }
+
+                Project lastSeenProject;
+                if (idTolastSeenProject.TryGetValue(project.Id, out lastSeenProject) && lastSeenProject == project)
+                {
+                    // We already saw this project.  No need to do anything;
+                    return;
                 }
 
-                ReportCount();
-                return SpecializedTasks.EmptyTask;
+                await UpdateReferencesAync(project, cancellationToken).ConfigureAwait(false);
+
+                var version = await project.GetSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+                ProjectInfo projectInfo;
+                if (!_projectToInfo.TryGetValue(project.Id, out projectInfo) || projectInfo.VersionStamp != version)
+                {
+                    var info = await SymbolTreeInfo.GetInfoForSourceAssemblyAsync(project, cancellationToken).ConfigureAwait(false);
+                    projectInfo = new ProjectInfo(version, info);
+                    _projectToInfo.AddOrUpdate(project.Id, projectInfo, (_1, _2) => projectInfo);
+                }
+
+                // Mark that we've completed processing this project.
+                idTolastSeenProject[project.Id] = project;
```

Comment:
Nothing ever removes this - does that mean that we strongly keep alive the last version of every project we ever come across?

Does this get cleaned up on closing a solution, or does it just leak?